### PR TITLE
reimplement Image Icon's in v7 & FontBook class

### PR DESCRIPTION
**Pull request by mggower**

### CODE REVIEW

File: src/renderers/webgl/index.ts [hunk 1]
```diff
@@ -99,14 +98,6 @@ export const defaultOptions = {
   dragInertia: 0.88
 }
 
-// TODO - make configurable
-export const MIN_LABEL_ZOOM = 0.25
```
Comment (mggower) [inline src/renderers/webgl/index.ts:103]: importing these in other files started a circle dependency so I moved them to `./utils`

**Review by jameslaneconkling — CHANGES_REQUESTED** [review]
In order for this to work w/ the image renderer, you'll need to track which remote assets are currently being loaded, and wait to generate the image until after they're rendered. otherwise image exports will be missing icons/labels. also, need to cancel the load task in the event that the node or renderer is destroyed. in v6 I managed this via a callback util called `Async`.

It's possible that in v7, pixi can now handle this natively via https://pixijs.com/guides/components/assets. Not sure...

**Review by jameslaneconkling — COMMENTED** [review]
Lot going on here, so if any of this sounds wrong to you, tell me, definitely not completely confident in everything below.

One source of bugs early on in the previous version was that promises can't resolve synchronously and also can't be cancelled.

Because they can't resolve synchronously, if you create a cache for something like a BitmapFont or image texture, and a getter function returns a promise that immediately resolves in case of a cache hit, the getter will still resolve in the next tick, not synchronously in the same tick (this is by design in the promise spec, and for the life of me I don't know why). This is usually bad, b/c any logic in the `then` clause won't run until after the current render completes w/ the call to `this.app.render()`.

Because promises can't be cancelled, there's a potential race condition where if you request a texture be created asynchronously, then tear down the Trellis instance before the texture is created, when the promise resolves and the pixi app has been destroyed, trying to create that texture will both throw an error and leak memory.

A final bug I kept running into: the image renderer needs to first render the scene, and if there are remote assets still being loaded, wait for them to load and be rendered before rendering the image.

It's possible to work around these limitations: for the first, have a synchronous cache check that you always use before loading anything. For the second, always check if the app has been destroyed before doing anything pixi related in a `then` clause. And for the third, check the loading status on all objects. But ultimately this was pretty brittle, so in the previous version I created the [Loader](https://github.com/sayari-analytics/trellis/blob/master/src/renderers/webgl-v6/Loader.ts) abstraction that's backed by this [Async util](https://github.com/sayari-analytics/trellis/commit/8645fefca57af4f1c97c277f3e5db0616ccdafec#diff-39b2554fd18da165b59a6351b1aafff3714e2a80c1435f2de9706355b4d32351R143-R229). (it's kind of like a functional promise: it's a higher order function, not a class, and it's synchronous, and cancellable)

My suggestion here would be:
* create `webgl/loader/imageLoader.ts`, `webgl/loader/fontLoader.ts`, replacing `webgl/textures/font`, `webgl/textures/icons` (decoupling the loading/cacheing of external assets from the creation of image textures, and from the creation of icon or label objects).
* use the loaders something like [here](https://github.com/sayari-analytics/trellis/blob/master/src/renderers/webgl-v6/node.ts#L392-L404) before creating any objects (b/c the loaders are synchronous, the callback is called immediately if there's a cache hit).
* cancel loaders when deleting a node/edge (this will noop if all loaders have already resolved).
* once we implement the image renderer, use the loader singletons so the image renderer can track the number of assets still loading.

**Review by jameslaneconkling — COMMENTED** [review]
Also, might make sense to roll up #94 into this PR

**Comment by mggower** [comment]
added edge labels in this PR

**Review by jameslaneconkling — COMMENTED** [review]
Left some more suggestions. Overall general theme is that we should strive to keep the NodeRenderer/EdgeRenderer as simple as possible by pushing logic to the objects and textures. NodeRenderer/EdgeRenderers and the parent renderer are already the location of most of the complexity, so anything like asset loading and texture creation/caching/management that can be pushed to the object/texture classes could be a big win.

I was having a harder time reasoning about the behavior and perf implications of the loaders themselves. The things to confirm is that
* rendering a graph w/ 50k nodes/edges doesn't degrade performance in comparison to what's currently on master
* renderer behavior is the same regardless of whether or not a loader is used to load a remote asset (e.g. `font: 'sans-serif'` which uses a native font, and `font: 'roboto'` which requires loading a remote font, should both produce equivalent graphs)
* it's possible to know whether remote assets are currently being loaded (so that the image renderer knows to not capture the image until after assets have loaded), and possible to cancel all requests for assets (so that we don't try to create pixi objects after the renderer has been torn down).

File: src/renderers/webgl/loaders/FontLoader.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import { noop } from '../../../utils'
+import FontFaceObserver from 'fontfaceobserver'
+
+type LoadFontProps = Partial<Subscriber<true>> & {
+  fontFamily: string
+  fontWeight?: string | number
+  timeout?: number
+}
+
+const GENERIC_FONT_FAMILIES = new Set([
```
[the remark this answers is not in the document]
Comment (mggower) [reply src/renderers/webgl/loaders/FontLoader.ts:12]: I pulled this list from [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/font-family#syntax). I suppose based on this [browser compatability](https://developer.mozilla.org/en-US/docs/Web/CSS/font-family#browser_compatibility) chart, we should drop `math` and `system-ui` from the list.

File: src/renderers/webgl/node.ts [hunk 3]
```diff
@@ -178,35 +264,34 @@ export class NodeRenderer {
...
-          this.renderer.labelObjectManager.mount(this.label)
-          this.labelMounted = true
-        }
-      } else {
-        if (this.labelMounted) {
-          this.renderer.labelObjectManager.unmount(this.label)
-          this.labelMounted = false
-        }
+      const shouldMount = isVisible && this.renderer.zoom > MIN_LABEL_ZOOM
+      if (shouldMount && !this.labelMounted) {
```
[the remark this answers is not in the document]
Comment (mggower) [reply src/renderers/webgl/node.ts:268]: Yeah I agree on standardizing this approach. The fill is the only render object in the node renderer not currently tied to an object manager, so the additional `fillMounted` means less. 

- I can add the node fill to an object manager
- I have also been playing around with a pattern that allows us to read an object's `mounted` state from the object manager so we would be able to remove that additional `iconMounted`/`labelMounted` values from the node renderer. Something like:

```typescript
class ObjectManager<Object extends RenderObject> {
  // ...
  
  isMounted(object: Object) {
    const code = this.batch.get(object)
    return code === 0 || (code === undefined && object.mounted)
  }
}

// and then in the node renderer 
const labelMounted = this.renderer.labelObjectManager.isMounted(this.label)
```

File: src/renderers/webgl/textures/textIcon/TextIconCache.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import type { Renderer } from '../..'
```
[the remark this answers is not in the document]
Comment (mggower) [reply src/renderers/webgl/textures/textIcon/TextIconCache.ts:1]: Yeah, great idea. I am happy to implement

File: src/renderers/webgl/loaders/AssetLoader.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      return null
+    }
+  }
+
+  load({ url, resolve = noop, reject = noop }: LoadAssetProps) {
+    if (this.cache[url] === undefined) {
+      this.cache[url] = this.createPublisher(url)
+    }
+
+    return this.cache[url].subscribe({ resolve, reject })
```
Comment (jameslaneconkling) [inline src/renderers/webgl/loaders/AssetLoader.ts:23]: Pretty sure I'm tracking the code path here correctly. But to confirm, if there's a cache hit, `this.cache[url].subscribe` will invoke `resolve` synchronously, yeah?

**Review by jameslaneconkling — APPROVED** [review]

**Pull request merged**